### PR TITLE
8075915: The eight controls without black backgrounds with WinLAF & GTK LAF & Nimbus LAF

### PR DESCRIPTION
**Pull request by azvegint**
This manual test was written for Metal L&F, so now it is set explicitly.
Test is also refactored to use JFrame instead of Applet.

The test is not in `ProblemList.txt`.
It still fails on JDK 1.5 (after removing all syntactic sugar).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8075915](https://bugs.openjdk.java.net/browse/JDK-8075915): The eight controls without black backgrounds with WinLAF & GTK LAF & Nimbus LAF

### Reviewers
 * [Prasanta Sadhukhan](https://openjdk.java.net/census#psadhukhan) (@prsadhuk - **Reviewer**) ⚠️ Review applies to f804ee1961fd7a44eaa0bb47e018140c0aade442
 * [Alexander Zuev](https://openjdk.java.net/census#kizune) (@azuev-java - **Reviewer**) ⚠️ Review applies to 86c154a2af6812136a1ef8dbf842a0284ebaff70
 * [Alexey Ivanov](https://openjdk.java.net/census#aivanov) (@aivanov-jdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3474/head:pull/3474` \
`$ git checkout pull/3474`

Update a local copy of the PR: \
`$ git checkout pull/3474` \
`$ git pull https://git.openjdk.java.net/jdk pull/3474/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3474`

View PR using the GUI difftool: \
`$ git pr show -t 3474`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3474.diff">https://git.openjdk.java.net/jdk/pull/3474.diff</a>

</details>

### CODE REVIEW

File: test/jdk/javax/swing/JCheckBox/4449413/bug4449413.java [hunk 1]
```diff
@@ -25,87 +25,141 @@
...
+    public static void main(String[] args) throws Exception {
+        UIManager.setLookAndFeel("javax.swing.plaf.metal.MetalLookAndFeel");
+
+        SwingUtilities.invokeLater(() -> new bug4449413().createAndShowGUI());
+
+        boolean timeoutHappened = !latch.await(2, TimeUnit.MINUTES);
+
+        if (timeoutHappened || failed) {
+            throw new RuntimeException("Test failed!");
+        }
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JCheckBox/4449413/bug4449413.java:87]: Shall the test dispose of the frame if it times out?

Comment (azvegint) [reply test/jdk/javax/swing/JCheckBox/4449413/bug4449413.java]: Actually `jtreg` handles it. But I've updated the test to support standalone mode.

**Comment by prsadhuk** [comment]
image

I see RadioButton and RBMenuItem check mark is not visible...so probably there is still a product bug somewhere...I guess we need to fix that as well when we fix this test..

**Comment by azvegint** [comment]
> I see RadioButton and RBMenuItem check mark is not visible...so probably there is still a product bug somewhere...I guess we need to fix that as well when we fix this test..

Actually it is clearly visible. You can click on it to see the difference between checked/unchecked state.
Probably you are referring to a circle shape of  "check mark", but that is correct shape.

Here is a comparison of JDK17 vs JDK5 (all components are in selected state).

JDK17
image
JDK5
image

JDK17
image
JDK5
image

JDK5 clearly failed this test.

**Comment by prsadhuk** [comment]
I am referring to check mark only, not the shape...I could see the white visible selected checkmark(circular) for RadioButton when "Ocean Theme" is not selected but when "Ocean Theme" is selected, black checkmark over black background is not visible. 

Also, the original bug 4449413 does not talk about Metal. It seems the issue is seen only in Metal but that should not prevent us from running in all L&Fs.

**Comment by azvegint** [comment]
> I am referring to check mark only, not the shape...I could see the white visible selected checkmark(circular) for RadioButton when "Ocean Theme" is not selected but when "Ocean Theme" is selected, black checkmark over black background is not visible.
>
This black check mark does have a light colored halo around which helps to see it.

Here is an example of radio button usage in group:
image
I definitely can see which one is selected. So I do not see issue here.

> Also, the original bug 4449413 does not talk about Metal. It seems the issue is seen only in Metal but that should not prevent us from running in all L&Fs.

Updated to allow it to run with other L&Fs.

Please see below screenshots with other L&Fs applied. Some of them controversial, but I prefer to file a new bugs for them if any.

image

image

image

image

image
(Labels for disabled text is not visible, but MotifLookAndFeel is marked for removal, so probably it is not worth to fix it.)

File: test/jdk/javax/swing/JCheckBox/4449413/bug4449413.java [hunk 2]
```diff
@@ -25,87 +25,170 @@
...
-        }
+        AbstractButton b = switch (type) {
+            case 0 -> new JRadioButton("RadioButton");
+            case 1 -> new JCheckBox("CheckBox");
+            case 2 -> new JRadioButtonMenuItem("RBMenuItem");
+            case 3 -> new JCheckBoxMenuItem("CBMenuItem");
+            default -> throw new IllegalArgumentException("type should be in range of 0..3");
+        };
+
+        b.setOpaque(true);
```
Comment (prsadhuk) [inline test/jdk/javax/swing/JCheckBox/4449413/bug4449413.java:191]: Any reason why this setOpaque was added? I am of the opinion that it should get controlled either by LookAndFeel.installProperty or by component's constructor.

Comment (azvegint) [reply test/jdk/javax/swing/JCheckBox/4449413/bug4449413.java:191]: The reason is in `JComponent#setBackground()` [javadoc](https://docs.oracle.com/en/java/javase/15/docs/api/java.desktop/javax/swing/JComponent.html#setBackground(java.awt.Color))



> Sets the background color of this component. The background color is used only if the component is opaque, and only by subclasses of JComponent or ComponentUI implementations. Direct subclasses of JComponent must override paintComponent to honor this property.
> 
> It is up to the look and feel to honor this property, some may choose to ignore it.

**Review by prsadhuk — APPROVED** [review]

**Review by azuev-java — APPROVED** [review]

File: test/jdk/javax/swing/JCheckBox/4449413/bug4449413.java [hunk 3]
```diff
@@ -25,87 +25,170 @@
...
+    public static void main(String[] args) throws Exception {
+        SwingUtilities.invokeLater(() -> {
+            instance = new bug4449413();
+            instance.createAndShowGUI();
+        });
+
+        boolean timeoutHappened = !latch.await(2, TimeUnit.MINUTES);
+        if (instance != null) {
+            instance.dispose();
+        }
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JCheckBox/4449413/bug4449413.java:96]: This should be wrapped into `invokeAndWait`.

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

File: test/jdk/javax/swing/JCheckBox/4449413/bug4449413.java (hunk 3 above)
Comment (azvegint) [reply test/jdk/javax/swing/JCheckBox/4449413/bug4449413.java]: Sure, fixed.

**Comment by azvegint** [comment]
I've also replaced text blocks with string literals to ease backporting.

**Review by aivanov-jdk — APPROVED** [review]

**Pull request closed without merging**